mountpoint-s3/src/inode.rs

```diff
@@ -439,6 +439,87 @@ impl Superblock {
 
         Ok(())
     }
+
+    /// Unlink the entry described by `parent_ino` and `name`.
+    ///
+    /// If the entry exists, delete it from S3 and the superblock.
+    ///
+    /// We know that the Linux Kernel's VFS will lock both the parent and child,
+    /// so we can safely ignore concurrent operations within the same Mountpoint process to the file and its parent.
+    /// See: https://www.kernel.org/doc/html/next/filesystems/directory-locking.html
+    pub async fn unlink<OC: ObjectClient>(
+        &self,
+        client: &OC,
+        parent_ino: InodeNo,
+        name: &OsStr,
+    ) -> Result<(), InodeError> {
+        let parent = self.inner.get(parent_ino)?;
+        let LookedUp { inode, .. } = self.lookup(client, parent_ino, name).await?;
+
+        if inode.kind() == InodeKind::Directory {
+            return Err(InodeError::IsDirectory(inode.ino()));
+        }
+
+        let write_status = {
+            let inode_state = inode.get_inode_state()?;
+            inode_state.write_status
+        };
+
+        match write_status {
+            WriteStatus::LocalUnopened | WriteStatus::LocalOpen => {
+                // In the future, we may permit `unlink` and cancel any in-flight uploads.
+                error!(
+                    parent = parent_ino,
+                    ?name,
+                    "unlink called on local file, unlink not supported until write is complete",
+                );
+                return Err(InodeError::UnlinkNotPermittedWhileWriting(inode.ino()));
+            }
+            WriteStatus::Remote => {
```

Comment:
What should be the paraemters like WriteStatus for `InodeState` once the Inode is unlinked? Or we dont really care? Like I get that lookup will not show up these files while create. But if a new object is created by any other client remotely, and `update_from_remote()` is called, it will still see the InodeStat as a `InodeKind::File` and `WriteStatus::Remote`.

Reply:
For the `WriteStatus`, I didn't end up updating this as the risk felt negligible.

If the Kernel still has the Inode number, it can still do things like `getattr` or `open`, however it feels OK for now to allow those to return. Maybe we should block.

What I can do is look into adding unit tests for those, as you suggested, to document how it works today and note the assumptions we're making / what we're OK with.